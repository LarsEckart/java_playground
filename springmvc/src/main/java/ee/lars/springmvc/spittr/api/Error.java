package ee.lars.springmvc.spittr.api;

public class Error {

    private final int code;
    private final String message;

    public Error(int code, String message) {
        this.code = code;
        this.message = message;
    }

    public int getCode() {
        return this.code;
    }

    public String getMessage() {
        return this.message;
    }
}