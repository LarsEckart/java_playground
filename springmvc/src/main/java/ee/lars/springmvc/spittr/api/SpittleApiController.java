package ee.lars.springmvc.spittr.api;

import ee.lars.springmvc.spittr.Spittle;
import ee.lars.springmvc.spittr.data.SpittleRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/spittles")
public class SpittleApiController {

    private static final String MAX_LONG_AS_STRING = "9223372036854775807";

    private final SpittleRepository spittleRepository;

    @Autowired
    public SpittleApiController(SpittleRepository spittleRepository) {
        this.spittleRepository = spittleRepository;
    }

    @RequestMapping(method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
    public List<Spittle> spittles(
            @RequestParam(value = "max", defaultValue = MAX_LONG_AS_STRING) long max,
            @RequestParam(value = "count", defaultValue = "20") int count)
    {
        return this.spittleRepository.findSpittles(max, count);
    }

    @RequestMapping(method = RequestMethod.POST, consumes = "application/json")
    public Spittle saveSpittle(@RequestBody Spittle spittle) {
        return spittleRepository.save(spittle);
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.GET)
    public ResponseEntity<?> spittleById(@PathVariable long id) {
        Spittle spittle = this.spittleRepository.findOne(id);
        if (spittle == null) {
            Error error = new Error(4, "Spittle [" + id + "] not found");
            return new ResponseEntity<Error>(error, HttpStatus.NOT_FOUND);
        }
        return new ResponseEntity<Spittle>(spittle, HttpStatus.OK);
    }
}
