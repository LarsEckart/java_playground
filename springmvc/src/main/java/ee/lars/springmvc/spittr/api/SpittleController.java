package ee.lars.springmvc.spittr.api;

import ee.lars.springmvc.spittr.Spittle;
import ee.lars.springmvc.spittr.data.SpittleRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;

import java.util.List;

@Controller
@RequestMapping("/spittles")
public class SpittleController {

    private static final String MAX_LONG_AS_STRING = "9223372036854775807";

    private final SpittleRepository spittleRepository;

    @Autowired
    public SpittleController(SpittleRepository spittleRepository) {
        this.spittleRepository = spittleRepository;
    }

    @RequestMapping(method = RequestMethod.GET)
    public List<Spittle> spittles(
            @RequestParam(value = "max", defaultValue = MAX_LONG_AS_STRING) long max,
            @RequestParam(value = "count", defaultValue = "20") int count)
    {
        return this.spittleRepository.findSpittles(max, count);
    }
}
