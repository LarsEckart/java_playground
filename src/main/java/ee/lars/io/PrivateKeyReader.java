package ee.lars.io;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.PrivateKey;
import java.security.spec.EncodedKeySpec;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.PKCS8EncodedKeySpec;

public class PrivateKeyReader {

    /**
     * Reads a private key file which is in PKCS#8 format.
     * For example create a private key via
     * openssl genrsa -out private_key.pem 2048
     * and then convert it via
     * openssl pkcs8 -topk8 -inform PEM -outform DER -in private_key.pem -out private_key.der -nocrypt
     */
    public static PrivateKey get(String filename)
            throws IOException, NoSuchAlgorithmException, InvalidKeySpecException
    {
        byte[] keyBytes = Files.readAllBytes(Paths.get(filename));

        EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(keyBytes);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        final PrivateKey privateKey = keyFactory.generatePrivate(keySpec);
        return privateKey;
    }
}
