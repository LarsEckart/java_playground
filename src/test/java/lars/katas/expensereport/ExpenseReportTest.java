package lars.katas.expensereport;

import com.github.larseckart.tcr.FastTestCommitRevertMainExtension;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.util.List;
import org.approvaltests.Approvals;
import org.approvaltests.combinations.CombinationApprovals;
import org.approvaltests.core.Options;
import org.approvaltests.scrubbers.DateScrubber;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(FastTestCommitRevertMainExtension.class)
class ExpenseReportTest {

  @Test
  void approval() {
    ExpenseType[] expenseType = new ExpenseType[]{ExpenseType.BREAKFAST, ExpenseType.DINNER, ExpenseType.CAR_RENTAL};
    Integer[] amount = new Integer[]{0, 1, 999, 1000, 1001, 4999, 5000, 5001};

    CombinationApprovals.verifyAllCombinations(
        this::createReport, expenseType, amount, new Options(DateScrubber.getScrubberFor("Wed Nov 17 22:28:33 EET 2021")));
  }

  private String createReport(ExpenseType expenseType, int amount) {
    final PrintStream systemOut = System.out;
    try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
      System.setOut(new PrintStream(baos, true, StandardCharsets.UTF_8));

      run(expenseType, amount);

      return baos.toString();
    } catch (IOException e) {
      throw new RuntimeException(e);
    } finally {
      System.setOut(systemOut);
    }
  }

  private void run(ExpenseType expenseType, int amount) {
    ExpenseReport expenseReport = new ExpenseReport();

    Expense expense = new Expense();
    expense.type = expenseType;
    expense.amount = amount;

    expenseReport.printReport(List.of(expense));
  }
}
