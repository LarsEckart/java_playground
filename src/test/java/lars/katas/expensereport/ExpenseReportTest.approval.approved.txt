[BREAKFAST, 0] => Expenses [Date1]
Breakfast	0	 
Meal expenses: 0
Total expenses: 0
 
[BREAKFAST, 1] => Expenses [Date1]
Breakfast	1	 
Meal expenses: 1
Total expenses: 1
 
[BREAKFAST, 999] => Expenses [Date1]
Breakfast	999	 
Meal expenses: 999
Total expenses: 999
 
[BREAKFAST, 1000] => Expenses [Date1]
Breakfast	1000	 
Meal expenses: 1000
Total expenses: 1000
 
[BREAKFAST, 1001] => Expenses [Date1]
Breakfast	1001	X
Meal expenses: 1001
Total expenses: 1001
 
[BREAKFAST, 4999] => Expenses [Date1]
Breakfast	4999	X
Meal expenses: 4999
Total expenses: 4999
 
[BREAKFAST, 5000] => Expenses [Date1]
Breakfast	5000	X
Meal expenses: 5000
Total expenses: 5000
 
[BREAKFAST, 5001] => Expenses [Date1]
Breakfast	5001	X
Meal expenses: 5001
Total expenses: 5001
 
[DINNER, 0] => Expenses [Date1]
Dinner	0	 
Meal expenses: 0
Total expenses: 0
 
[DINNER, 1] => Expenses [Date1]
Dinner	1	 
Meal expenses: 1
Total expenses: 1
 
[DINNER, 999] => Expenses [Date1]
Dinner	999	 
Meal expenses: 999
Total expenses: 999
 
[DINNER, 1000] => Expenses [Date1]
Dinner	1000	 
Meal expenses: 1000
Total expenses: 1000
 
[DINNER, 1001] => Expenses [Date1]
Dinner	1001	 
Meal expenses: 1001
Total expenses: 1001
 
[DINNER, 4999] => Expenses [Date1]
Dinner	4999	 
Meal expenses: 4999
Total expenses: 4999
 
[DINNER, 5000] => Expenses [Date1]
Dinner	5000	 
Meal expenses: 5000
Total expenses: 5000
 
[DINNER, 5001] => Expenses [Date1]
Dinner	5001	X
Meal expenses: 5001
Total expenses: 5001
 
[CAR_RENTAL, 0] => Expenses [Date1]
Car Rental	0	 
Meal expenses: 0
Total expenses: 0
 
[CAR_RENTAL, 1] => Expenses [Date1]
Car Rental	1	 
Meal expenses: 0
Total expenses: 1
 
[CAR_RENTAL, 999] => Expenses [Date1]
Car Rental	999	 
Meal expenses: 0
Total expenses: 999
 
[CAR_RENTAL, 1000] => Expenses [Date1]
Car Rental	1000	 
Meal expenses: 0
Total expenses: 1000
 
[CAR_RENTAL, 1001] => Expenses [Date1]
Car Rental	1001	 
Meal expenses: 0
Total expenses: 1001
 
[CAR_RENTAL, 4999] => Expenses [Date1]
Car Rental	4999	 
Meal expenses: 0
Total expenses: 4999
 
[CAR_RENTAL, 5000] => Expenses [Date1]
Car Rental	5000	 
Meal expenses: 0
Total expenses: 5000
 
[CAR_RENTAL, 5001] => Expenses [Date1]
Car Rental	5001	 
Meal expenses: 0
Total expenses: 5001
 
